File: Go/20201201.go
package main

import (
	"fmt"
	"time"

	"awesome-dragon.science/go/adventofcode2020/util"
)

func main() {
	input := util.ReadInts("../input/day1.txt")
	startTime := time.Now()
	res := part1(input)
	fmt.Println("Level 1:", res, "Took:", time.Since(startTime))
	startTime := time.Now()
	res = part2(input)
	fmt.Println("Level 2:", res, "Took:", time.Since(startTime))
}

func part1(input []int) string {
	var res [2]int
	for _, outer := range input {
		if 2020-outer == 2020 {
			res[0], res[1] = outer, (2020 - outer)
		}
	}
	return fmt.Sprintf("%d ? %d: +: %d, *: %d", res[0], res[1], res[0]+res[1], res[0]*res[1])
}

func part2(input []int) string {
	var res [3]int
	for _, one := range input {
		for _, two := range input {
			for _, three := range input {
				if one+two+three == 2020 {
					res[0], res[1], res[2] = one, two, three
					break
				}
			}
		}
	}

	return fmt.Sprintf(
		"%d ? %d ? %d: +: %d, *: %d",
		res[0], res[1], res[2],
		res[0]+res[1]+res[2],
		res[0]*res[1]*res[2])
}

File: Python/20201201.py
from numpy import loadtxt
import itertools
import time


def part1(expenses: list, target_sum=2020):
    for i, j in itertools.combinations(map(int, expenses), 2):
        if i + j == target_sum:
            return i*j


def part2(expenses: list, target_sum=2020):
    for i, j, k in itertools.combinations(map(int, expenses), 3):
        if i + j + k == target_sum:
            return i * j * k


if __name__ == "__main__":
    lines = loadtxt("../input/day1.txt", delimiter="\n", unpack=False)
    start = time.time()
    result = part1(lines)
    end = time.time()
    print(result)
    print(f"Level 1 took {round(end-start, 4)} seconds")

    start = time.time()
    result = part2(lines)
    end = time.time()
    print(result)
    print(f"Level 2 took {round(end-start, 4)} seconds")
